pkg/email/send_email.go

```diff
@@ -11,30 +14,35 @@ type Emailer struct {
 	Limiter *rate.Limiter
 }
 
+var (
+	ErrRateLimitExceeded = errors.New("rate limit exceeded")
+)
+
 const (
 	defaultRateLimit = 1.0 / 60.0
 	defaultBurstRate = 10
 )
 
 type EmailParams struct {
-	SMTPHost  string
-	Port      int
-	Username  string
-	Password  string
-	Sender    string
-	RateLimit float64
-	Burst     int
+	SMTPHost   string
+	Port       int
+	Username   string
+	Password   string
+	Sender     string
+	LimitEvery float64
+	Burst      int
 }
 
 func NewEmailer(e EmailParams) Emailer {
 	d := gomail.NewDialer(e.SMTPHost, e.Port, e.Username, e.Password)
-	if e.RateLimit == 0 {
-		e.RateLimit = defaultRateLimit
+	if e.LimitEvery == 0 {
+		e.LimitEvery = defaultRateLimit
 	}
 	if e.Burst == 0 {
 		e.Burst = defaultBurstRate
 	}
-	l := rate.NewLimiter(rate.Limit(e.RateLimit), e.Burst)
+	limit := float64(time.Second) / float64(e.LimitEvery)
+	l := rate.NewLimiter(rate.Limit(limit), e.Burst)
```

Comment:
If there was already discussion about the rate limit, but my 2cent:
1. https://pkg.go.dev/golang.org/x/time/rate#Every - based on the configuration, this is the method you are looking for.
2. based on the current implementation, if both of us asks for password reset - at the same time, my request will fail. We don't tell the user about our failures. so, I will not get any email.  So as I see it, we will require to tell the user we failed, so it can retry (not something I like).  Have different rate limit, one that holds different configuration for the global (all emails) and per email.

Reply:
(1) is really true, thanks!

(2) -- this is _exactly_ what the burst is for.  We can send _burst_ emails at once when we do not suspect an attack.  So not a problem when not under attack.  We might additionally report an informative error to users (of course).

Reply:
Thanks @arielshaqed this is what I was looking for - item 2.